cilium_eip_no_masquerade_agent/src/main.rs

```diff
@@ -0,0 +1,216 @@
+use std::net::Ipv4Addr;
+use std::sync::Arc;
+use std::time::Duration;
+
+use futures::{future, StreamExt, TryStream, TryStreamExt};
+use ipnetwork::Ipv4Network;
+use k8s_openapi::api::core::v1::Pod;
+use kube::api::{Api, ListParams};
+use kube::{Client, ResourceExt};
+use kube_runtime::controller::{Action, Controller};
+use kube_runtime::finalizer::{finalizer, Event};
+use rand::{thread_rng, Rng};
+use rtnetlink::packet::rule::Nla;
+use rtnetlink::packet::RuleMessage;
+use rtnetlink::{new_connection, Handle, IpVersion, RuleHandle};
+use tracing::{debug, event, info, instrument, Level};
+
+use eip_operator_shared::{run_with_tracing, Error};
+
+const MANAGE_EIP_LABEL: &str = "eip.materialize.cloud/manage";
+const POD_FINALIZER_NAME: &str = "eip.materialize.cloud/cilium_no_masquerade_rule";
+
+struct ContextData {
+    k8s_client: Client,
+    handle: Handle,
+    vpc_cidr: Ipv4Network,
+}
+
+impl ContextData {
+    fn new(k8s_client: Client, handle: Handle, vpc_cidr: Ipv4Network) -> ContextData {
+        ContextData {
+            k8s_client,
+            handle,
+            vpc_cidr,
+        }
+    }
+}
+
+async fn filter_pod_rules(
+    rules: impl TryStream<Ok = RuleMessage, Error = rtnetlink::Error>,
+    pod_ip: Ipv4Addr,
+) -> Result<Vec<RuleMessage>, rtnetlink::Error> {
+    rules
+        .try_filter(|rule| {
+            future::ready(
+                rule.header.src_len == 32
+                    && rule.nlas.contains(&Nla::Source(pod_ip.octets().to_vec())),
+            )
+        })
+        .collect::<Vec<Result<RuleMessage, rtnetlink::Error>>>()
+        .await
+        .into_iter()
+        .collect()
+}
+
+#[instrument(skip(rule_handle, pod), err)]
+async fn apply_pod(
+    rule_handle: RuleHandle,
+    vpc_cidr: Ipv4Network,
+    pod: Arc<Pod>,
+) -> Result<Action, Error> {
+    let pod_name = pod.metadata.name.as_ref().ok_or(Error::MissingPodName)?;
+    event!(Level::INFO, pod_name = %pod_name, "Applying pod.");
+    let pod_ip: Ipv4Addr = pod
+        .status
+        .as_ref()
+        .ok_or(Error::MissingPodIp)?
+        .pod_ip
+        .as_ref()
+        .ok_or(Error::MissingPodIp)?
+        .parse()?;
+    let rules = rule_handle.get(IpVersion::V4).execute();
+    let pod_rules = filter_pod_rules(rules, pod_ip).await?;
+
+    // Find table from Cilium's installed rule
+    let table = pod_rules
+        .iter()
+        .find_map(|rule| {
+            if rule.header.dst_len == vpc_cidr.prefix()
+                && rule
+                    .nlas
+                    .contains(&Nla::Destination(vpc_cidr.network().octets().to_vec()))
+            {
+                Some(rule.header.table)
+            } else {
+                None
+            }
+        })
+        .ok_or(Error::CiliumRuleNotFound)?;
```

Comment:
This causes it to get rescheduled later, right?

Reply:
Yes, it gets requeued for a random time between 4-8 seconds later.